op-node/rollup/derive/frame_queue_test.go

```diff
@@ -0,0 +1,154 @@
+package derive
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log/slog"
+	"testing"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/derive/mocks"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPruneFrameQueue(t *testing.T) {
+	for _, tt := range []struct {
+		desc string
+		fs   []testFrame
+		exp  []testFrame
+	}{
+		{
+			desc: "empty",
+			fs:   []testFrame{},
+			exp:  []testFrame{},
+		},
+		{
+			desc: "one",
+			fs:   []testFrame{"a:2:"},
+			exp:  []testFrame{"a:2:"},
+		},
+		{
+			desc: "one-last",
+			fs:   []testFrame{"a:2:!"},
+			exp:  []testFrame{"a:2:!"},
+		},
+		{
+			desc: "last-new",
+			fs:   []testFrame{"a:2:!", "b:0:"},
+			exp:  []testFrame{"a:2:!", "b:0:"},
+		},
+		{
+			desc: "last-ooo",
+			fs:   []testFrame{"a:2:!", "b:1:"},
+			exp:  []testFrame{"a:2:!"},
+		},
+		{
+			desc: "middle-lastooo",
+			fs:   []testFrame{"b:1:", "a:2:!"},
+			exp:  []testFrame{"b:1:"},
+		},
+		{
+			desc: "middle-first",
+			fs:   []testFrame{"b:1:", "a:0:"},
+			exp:  []testFrame{"a:0:"},
+		},
+		{
+			desc: "last-first",
+			fs:   []testFrame{"b:1:!", "a:0:"},
+			exp:  []testFrame{"b:1:!", "a:0:"},
+		},
+		{
+			desc: "last-ooo",
+			fs:   []testFrame{"b:1:!", "b:2:"},
+			exp:  []testFrame{"b:1:!"},
+		},
+		{
+			desc: "ooo",
+			fs:   []testFrame{"b:1:", "b:3:"},
+			exp:  []testFrame{"b:1:"},
+		},
```

Comment:
Probably worth a test where the out of order frame is received then the missed one then the out of order one in the right place (ie it is resubmitted):

```
fs : []testFrame{"b:1:", "b:3:!", "b:2:", "b:3:!"}
exp: []testFrame{"b:1:", "b:2:", "b:3:!"}

Reply:
good idea, added.